docs/source/install.rst

```diff
@@ -170,6 +170,6 @@ Building and installing UCX-Py can be done via `pip install`. For example:
     conda activate ucx
     git clone https://github.com/rapidsai/ucx-py.git
     cd ucx-py
-    pip install -v .
+    pip install -v -C rapidsai.matrix-entry="cuda=12.2" .
```

Comment:
Rather than always including this, should the default development instructions not enable CUDA, and therefore not specify a matrix entry? This assumes we make the other change I suggest above.

Reply:
I think this is a question for @pentschev ?

These docs are under a section that says it's assumed you're building on a system with CUDA.

https://github.com/rapidsai/ucx-py/blob/7b70211452d73ef8a0486b0b8ffa04777ee6fef4/docs/source/install.rst#L50-L53

I'm happy to change this PR either way, but I don't feel qualified to make the call about whether or not the CUDA-enabled build should be the default.

I DO think that we should continue specifying this flag (and therefore depending on the `libucx` package) in docs builds here. Since we've seen that those docs builds end up also carrying some test coverage to catch works-without-CUDA types of issues (https://github.com/rapidsai/ucxx/issues/231).

So @pentschev , the specific question for you is, which of these would you prefer?

* `pip install .` works with 0 other flags, but requires a system installation of UCX
* `pip install .` requires passing `-C rapidsai.matrix-entry="cuda12.2"` or similar, but in exchange the locally-built package depends on `libucx` wheels (so no system version required)

Reply:
I just pushed https://github.com/rapidsai/ucx-py/pull/1048/commits/7b675a48f80c66d84e6230d0ca19b246e866816a implementing this, by the way, so you can look at the diff and see what it'd look like.

Reply:
The docs assume a CUDA system because of the `--with-cuda=$CUDA_HOME` build flag, which is our primary use case, if that flag is omitted then it should work on non-CUDA systems as well, but we don't document that at the moment.

> the specific question for you is, which of these would you prefer?
> 
> * `pip install .` works with 0 other flags, but requires a system installation of UCX
> * `pip install .` requires passing `-C rapidsai.matrix-entry="cuda12.2"` or similar, but in exchange the locally-built package depends on `libucx` wheels (so no system version required)

Can I have both? 🙂 

Ideally, I would like both to work, but given your phrasing it seems this is not possible. So let's assume you have UCX installed on the system and `libucx` installed, what would happen then both with and without `-C rapidsai.matrix-entry="cuda12.2"`?

Reply:
> *assume you have `libucx` installed*

This question is about whether or not `pip install .` pulls that dependency in **for you**, not about what happens if it happens to already be installed.

If you've already done something like the following:

```shell
apt-get install openucx
pip install libucx-cu12
```

Then `import ucp` will call `libucx.load_library()`, which should find that system installation of `libucs`, `libucm`, etc., but will fall back to the `libucx`-bundled ones if finding them fails for some reason.

Ok so what's the point of `-C rapidsai.matrix-entry`? `libucx` isn't distributed as the name `libucx`.... you have to pick `libucx-cu11` (CUDA 11) or `libucx-cu12` (CUDA 12). That selection of major CUDA version can happen automatically in `rapids-build-backend`, but that detection is being turned off here via the use of `disable-cuda = true`, to support source installation on non-CUDA systems by default.

Another option we could pursue, if you want, is to choose a specific CUDA verrsion as the default one for source installations by e.g. recording `libucx-cu12` as a dependency in `pyproject.toml`. Then `pip install .` would "just work" with no additional flags on a system with or without CUDA available and with or without a system installation of UCX. But anyone using CUDA 11 would have to pass `-C rapidsai.matrix-entry="cuda11.8"` or similar, or they'd get a `libucx-cu12` (built against CUDA 12), which might lead to runtime issues.

Reply:
Whenever you have time @pentschev , this is ready for another review.

I've taken @vyasr 's suggestions in this thread + https://github.com/rapidsai/ucx-py/pull/1048#discussion_r1651370374 and changes this so that there is no `libucx` dependency in `pyproject.toml` by default.

In the current state of this PR, wheels published to PyPI would still have that dependency, but running `pip install .` from the root of the repo would require that you provide UCX libraries some other way and should work with any of the following:

* building UCX yourself from source
  - *(like https://github.com/rapidsai/devcontainers/blob/branch-24.08/features/src/ucx/install.sh)*
* using an installation from another package manager
  - *(e.g. `apt-get install openucx` or `conda install ucx`)*
* installing the `libucx` wheels manually
  - *(e.g. `pip install libucx-cu{11,12}`)*

If you're ok with that being the development experience, then this PR could be merged as-is. The docs can continue to specify `pip install .` with 0 flags, like this:

https://github.com/rapidsai/ucx-py/blob/7b70211452d73ef8a0486b0b8ffa04777ee6fef4/docs/source/install.rst#L171-L173

If you want source installation to work without *requiring* a separate installation of UCX, then `ucx-py` built from source needs a `libucx` dependency.

And it could get it from any of the following:

* in the docs, instructing people to pass `-C rapidsai.matrix-entry="cuda=12.2"` (and similar for CUDA 11)
* putting a default matrix entry in `pyproject.toml` (e.g. either defaulting to CUDA 11 or CUDA 12)

Reply:
This looks great, I tested with system and conda installation of UCX and both worked as expected! Thanks a lot for pushing this through @jameslamb .

The only thing I would document, not just for a public reference of how it should be done but for my reference as well, is how to install the package with UCX wheels as well. Would `pip install --extra-index-url=https://pypi.nvidia.com libucx-cu1{1,2}; pip install -v .` work or is something more elaborate necessary?

Reply:
The above alone doesn't work (it still links to the system install), either with or with `-C rapidsai.matrix-entry="cuda=12.2"`, I suspect this is because the library is under `$CONDA_PREFIX/lib/python3.XY/site-packages/libucx` vs `$CONDA_PREFIX` when using the conda package. Is there a way to get the installer to find the `libucx` package?

Reply:
> The above alone doesn't work (it still links to the system install),

It's expected that even if `libucx` is installed, system libraries will be found first. That's intentional behavior.

See:

https://github.com/rapidsai/ucx-wheels/blob/ff1946193342bdc75c082c3d6aa153d8cdd400a9/python/libucx/libucx/load.py#L37-L39

If you want to force the use of `libucx`, even when there's another installation of the UCX libraries available, you'd have to do one of the following:

* remove those other installations (delete the files or at least move them to a path not checked by the loader/linker)
* use environment variable `LD_LIBRARY_PATH` (or maybe `LD_PRELOAD`, depending on the context) to force the ones from the `libucx` wheel to be found first

Knowing that, do you still want me to add docs explaining how to do that?

> `--extra-index-url https://pypi.nvidia.com`

By the way, `libucx` is up on public PyPI: https://pypi.org/project/libucx-cu12/

So I don't think the `--extra-index-url` should be required. The packages there on `pypi.org` are the real wheels, not stubs that redirect to the NVIDIA index. And in fact, we don't even have `libucx` published on`pypi.nvidia.com`: https://pypi.nvidia.com/

Reply:
> It's expected that even if libucx is installed, system libraries will be found first.

Ah right, I forgot about that.

> * remove those other installations (delete the files or at least move them to a path not checked by the loader/linker)

That is usually not possible, if it's a system install where you're not an admin you can't do it.

> * use environment variable `LD_LIBRARY_PATH` (or maybe `LD_PRELOAD`, depending on the context) to force the ones from the `libucx` wheel to be found first

But in that case `LD_LIBRARY_PATH` needs to always be set, not just when building. Would there be any other way of telling `pip install` to link to libraries in a specific path such as `$CONDA_PREFIX/lib/python3.XY/site-packages/libucx`? I don't think it's worth spending too much time on this, but I'm picking your brain in case there's a simple solution, if there isn't one then we can document the `LD_LIBRARY_PATH` solution and be done with it.

> By the way, `libucx` is up on public PyPI: https://pypi.org/project/libucx-cu12/
> 
> So I don't think the `--extra-index-url` should be required. The packages there on `pypi.org` are the real wheels, not stubs that redirect to the NVIDIA index. And in fact, we don't even have `libucx` published on`pypi.nvidia.com`: https://pypi.nvidia.com/

Ah right, my bad. Even simpler then. 😄

Reply:
No problem! There are lots of interconnected bits of this.

> Would there be any other way of telling `pip install` to link to libraries in a specific path such as `$CONDA_PREFIX/lib/python3.XY/site-packages/libucx`?

For as long as we want the default behavior to be that the libraries from the `libucx` wheels are only used if there are not any other installations of UCX around, I can't think of any other, "simple" ways to override that behavior besides those I listed above.

I think we should document and test the `LD_LIBRARY_PATH` approach (I can do that today), and not go further until we observe a use-case where managing the UCX installation via `pip` (e.g. being able to use `pip install libucx` to get a different version than the system install) is preferred and where the `LD_LIBRARY_PATH` approach is too cumbersome.

Reply:
> I think we should document and test the `LD_LIBRARY_PATH` approach (I can do that today), and not go further until we observe a use-case where managing the UCX installation via `pip` (e.g. being able to use `pip install libucx` to get a different version than the system install) is preferred and where the `LD_LIBRARY_PATH` approach is too cumbersome.

I'm absolutely of the same opinion, if there was a simple path that would be good, but it's not worth spending more time here. If you can document it, that would be great, I can also work on it and push to this branch if you prefer.

Reply:
Thanks for talking through it with!

Ok sure, you could do it and push to this branch. You know this project's documentation and the audience for it much better than I do 😊

Reply:
Done in https://github.com/rapidsai/ucx-py/pull/1048/commits/a39811cc14034fc4890086dadd8e5022dddcdd28. Would you mind having a look if everything seems correct to you @jameslamb ?

Also, I've accidentally committed [this local change](https://github.com/rapidsai/ucx-py/pull/1048/files#diff-4ce4b8bd5b44b77d2e86d1f995f3c9e47e819dcf799a12fcd1e974c494e67cc4L10-R10) that I was going to push to another PR. It's harmless and corrects the environment now, so if you don't mind let's leave it.

Reply:
I'd written up a long comment about how the docs should indicate that CPU-only is totally supported by `pip install ucx-py-cu{11,12}` .... then re-read what you wrote and saw that you already said exactly that 🤦🏻 

Anyway, in case it's interesting, here's how I tested that to convince myself:

<details><summary>how I tested that (click me)</summary>

Ran the following on my macOS laptop (so no possibility of accidentally finding CUDA):

```shell
docker run \
  --rm \
  -it python:3.11 \
  bash

# system-install UCX so we don't link to the the wheel one
apt-get update
apt-get install -y --no-install-recommends \
  libucx-dev

# install ucx-py-cu12 (which also pulls in libucx-cu12)
pip install --extra-index-url https://pypi.nvidia.com 'ucx-py-cu12>=0.38'

# import ucx-py and check where the loader found libraries
python -c "import ucp"
ldconfig -p | grep libuc
#	libuct.so.0 (libc6,AArch64) => /lib/aarch64-linux-gnu/libuct.so.0
#	libuct.so (libc6,AArch64) => /lib/aarch64-linux-gnu/libuct.so
#	libucs_signal.so.0 (libc6,AArch64) => /lib/aarch64-linux-gnu/libucs_signal.so.0
#	libucs_signal.so (libc6,AArch64) => /lib/aarch64-linux-gnu/libucs_signal.so
# ... etc., etc. ... it found them all

# write the demo code from https://github.com/rapidsai/ucx-py/blob/branch-0.39/docs/source/quickstart.rst
cat > server.py <<EOF
import asyncio
import time
import ucp
import numpy as np

n_bytes = 2**30
host = ucp.get_address(ifname='eth0')  # ethernet device name
port = 13337

async def send(ep):
    # recv buffer
    arr = np.empty(n_bytes, dtype='u1')
    await ep.recv(arr)
    assert np.count_nonzero(arr) == np.array(0, dtype=np.int64)
    print("Received NumPy array")

    # increment array and send back
    arr += 1
    print("Sending incremented NumPy array")
    await ep.send(arr)

    await ep.close()
    lf.close()

async def main():
    global lf
    lf = ucp.create_listener(send, port)

    while not lf.closed():
        await asyncio.sleep(0.1)

if __name__ == '__main__':
    asyncio.run(main())
EOF

cat > ./client.py <<EOF
import asyncio
import ucp
import numpy as np

port = 13337
n_bytes = 2**30

async def main():
    host = ucp.get_address(ifname='eth0')  # ethernet device name
    ep = await ucp.create_endpoint(host, port)
    msg = np.zeros(n_bytes, dtype='u1') # create some data to send

    # send message
    print("Send Original NumPy array")
    await ep.send(msg)  # send the real message

    # recv response
    print("Receive Incremented NumPy arrays")
    resp = np.empty_like(msg)
    await ep.recv(resp)  # receive the echo
    await ep.close()
    np.testing.assert_array_equal(msg + 1, resp)
    print("successfully received")

if __name__ == '__main__':
    asyncio.run(main())
EOF

# start the client and server
python ./server.py &
python ./client.py &
```

Saw output like this:

```text
Send Original NumPy array
Receive Incremented NumPy arrays
Received NumPy array
Sending incremented NumPy array
successfully received
```

</summary>

Maybe that'd be a helpful smoke test to add for https://github.com/rapidsai/ucxx/issues/231.

Anyway... everything you wrote looks correct and clear to me! I think this is good to merge. Thanks as always for your patience working through it with me 😊

Reply:
Smoke tests are always useful, my concern is how would that work in CI? Not sure if it's worth the trouble if we have to add another job at this time, from existing jobs it's probably relatively complicated to do so in a "safe" manner (i.e., ensuring linkage isn't resolved via the proper PyPI/conda UCX package).

I'm also +1 on merging this, I've approved the PR. Thanks so much James for the work here! 😄

Reply:
> *Not sure if it's worth the trouble*

For sure, this wasn't a proposal to do add a new job or anything **right now**.

Thanks! I'll merge this once CI passes.